test/csuite/wt8659_reconstruct_database_from_logs/main.c

```diff
@@ -0,0 +1,379 @@
+{
+    /*
+     * The full backup here is only needed for testing and comparison purposes. A normal incremental
+     * backup procedure would not include this.
+     */
+    printf("Taking full backup\n");
+    take_full_backup(home_live, home_full);
+
+    /*
+     * Taking the incremental backup also calls truncate to archive the log files, if the copies
+     * were successful. See that function for details on that call. The truncation only happens
+     * after we reconnected to the database.
+     */
+    printf("Taking incremental backup\n");
+    take_incr_backup(home_incr, after_reconnect);
+
+    /*
+     * Dump tables from the full backup and incremental backup databases, and compare the dumps.
+     */
+    printf("Dumping and comparing data\n");
+    testutil_check(compare_backups());
+    reset_dir(home_full);
+}
+
+/*
+ * usage --
+ *     Print out the command usage message.
+ */
+static void
+usage(void)
+{
+    fprintf(stderr, "Usage: %s [-w path] [-i number]\n", progname);
+    fprintf(stderr, "\t-w <path> optional path to the wt tool. Default is \"./wt\".\n");
+    fprintf(stderr, "\t-i <number> optional number of iterations to run. Default is 5.\n");
+
+    exit(EXIT_FAILURE);
+}
+
+/*
+ * parse_args --
+ *     Parse command line arguments.
+ */
+static void
+parse_args(int argc, char *argv[])
+{
+    int ch;
+
+    while ((ch = __wt_getopt(progname, argc, argv, "w:i:")) != EOF)
+        switch (ch) {
+        case 'w':
+            wt_tool_path = __wt_optarg;
+            if (access(wt_tool_path, F_OK) != 0) {
+                printf("Invalid path to WT tool: %s\n", __wt_optarg);
+                usage();
+            }
+            break;
+        case 'i':
+            iterations_num = atoi(__wt_optarg);
+            if (iterations_num <= 0) {
+                printf("Invalid iterations number: %s\n", __wt_optarg);
+                usage();
+            }
+            break;
+        default:
+            break;
+        }
+}
+
+/*
+ * main --
+ *     Test's entry point.
+ */
+int
+main(int argc, char *argv[])
+{
+    int i;
+
+    (void)testutil_set_progname(argv);
+    parse_args(argc, argv);
+
+    prepare_folders();
+
+    reopen_conn();
+    testutil_check(session->create(session, uri, table_config));
+
+    printf("Taking initial backup into incremental backup folder\n");
+    take_full_backup(home_live, home_incr);
+
+    for (i = 1; i <= iterations_num; i++) {
+        printf("==================================\n");
+        printf("Iteration %d:\n", i);
+        printf("==================================\n");
+
+        printf("Adding data\n");
+        add_work(i);
+        testutil_check(session->checkpoint(session, NULL));
+
+        /* Validate database against incremental backup. */
+        validate(false);
+
+        /* Reopen connection. */
+        reopen_conn();
+
+        /* Validate database again. */
+        validate(true);
+    }
+
+    cleanup();
+
+    return (EXIT_SUCCESS);
```

Comment:
Not specific to this line but overall:
This test is pretty much identical to `examples/c/ex_backup.c`. Since it is so similar, it is hard for me to see in the code what is different or new in this test. What is the added value in this test compared to that example?

Reply:
The main difference between the two is that in `wt8659_reconstruct_database_from_logs` at the end of each iteration we validate data, do a restart, and validate again. While `ex_backup` doesn't do a restart. Also, the way the backups are handled is a bit different. If you are interested here are the algorithms of both modules:

**ex_backup**
1. Create a working database `WT_HOME_LOG`
2. Crete a test table in `WT_HOME_LOG`
3. In `WT_HOME_LOG` add some data to the table
4. Take a full backup `WT_HOME_LOG` -> `WT_HOME_LOG_INCR.0` ... `WT_HOME_LOG_INCR.4`
5. Checkpoint `WT_HOME_LOG`
6. In `WT_HOME_LOG` add some data to the table and checkpoint it
7. Take a full backup `WT_HOME_LOG` -> `WT_HOME_LOG_FULL.<idx>`
8. Take an incremental backup with logs truncation `WT_HOME_LOG` -> `WT_HOME_LOG_INCR.<idx>` ... `WT_HOME_LOG_INCR.4`
9. Dump and compare the data in `WT_HOME_LOG_FULL.<idx>` vs `WT_HOME_LOG_INCR.<idx>`
10. Repeat iteration from step # 6

**wt8659_reconstruct_database_from_logs**
1. Create a working database in `WT_HOME_LOG`
2. Crete a test table in `WT_HOME_LOG`
3. Take an initial full backup `WT_HOME_LOG` -> `WT_HOME_LOG_INCR`
4. In `WT_HOME_LOG` add some data to the table and checkpoint it
5. Validate the database
    a. Take a full backup `WT_HOME_LOG` -> `WT_HOME_LOG_FULL`
    b. Take an incremental log-based backup without truncating logs `WT_HOME_LOG` -> `WT_HOME_LOG_INCR`
    c. Compare data in `WT_HOME_LOG_FULL` vs `WT_HOME_LOG_INCR`
        i. Copy `WT_HOME_LOG_INCR` into `WT_HOME_LOG_INCR_COPY`. This is required to avoid "Incremental backup after running recovery is not allowed" when we will do `wt dump`.
        ii. Dump the test table content from `WT_HOME_LOG_FULL` and `WT_HOME_LOG_INCR_COPY`.
        iii. Compare the dumps.
6. Reopen the connection to the `WT_HOME_LOG`.
7. Validate the database. This time with logs truncation in step # 5b.
8. End of iteration. The next one starts from step # 4.

Reply:
Thanks.